dask/order.py

```diff
@@ -202,125 +198,255 @@ def dependencies_key(x):
             StrComparable(x),
         )
 
+    def finish_now_key(x):
+        """ Determine the order of dependents that are ready to run and be released"""
+        return (-len(dependencies[x]), StrComparable(x))
+
+    # Computing this for all keys can sometimes be relatively expensive :(
```

Comment:
Here are the bencmarks for master against this PR

```
       before           after         ratio
     [278ad0a4]       [35567e48]
     <master>         <pull/5872/head>
+         289±4ms       1.04±0.02s     3.59  order.OrderFullLayers.time_order_full_layers((2, 10000))
+         267±2ms          385±6ms     1.44  order.OrderFullLayers.time_order_full_layers((10, 1000))
+         151±6ms          190±3ms     1.26  order.OrderLinearWithDanglers.time_order_linear_danglers((2, 10000))
+         173±4ms          210±8ms     1.21  order.OrderManySubgraphs.time_order_many_subgraphs((1, 9999))
+         204±6ms          247±7ms     1.21  order.OrderManySubgraphs.time_order_many_subgraphs((100, 99))
+         200±2ms          238±7ms     1.19  order.OrderManySubgraphs.time_order_many_subgraphs((10, 999))
+         184±2ms          216±2ms     1.17  order.OrderManySubgraphs.time_order_many_subgraphs((30, 303))
+         281±7ms          322±7ms     1.15  order.OrderLinalgSolves.time_order_linalg_solve
+         365±7ms         405±10ms     1.11  order.OrderFullLayers.time_order_full_layers((500, 2))
```

@eriknw do you have a sense for what percent of `partition_keys` we'll actually end up using? I suppose it depends on the dask graph...

I have a local version using functools.lru_cache, but it's slower (presumably because we're accessing every element).

Reply:
Yeah, I think we access pretty much everything.  We may not always access `partition_keys` for nodes that only have one dependent.  When I was profiling, I think the worst case was that creating `partition_keys` took about a quarter of the time.

I would guess that the one that is 3.59 times slower is about the absolute worst case scenario.  I think it's a pathologically bad case, since everything is the smallest for which we need to partition things into dicts of lists.  We specialize the case `len(deps) == 1`.  We may be able to specialize the case `len(deps) == 2` (or `< n` for small `n`).  We try to make `deps` as small as possible before processing, so adding additional fast paths for small `deps` could indeed be beneficial.  I don't have the time this week to look into this.

Overall, these benchmarks look good to me.  For ~10%-25% more processing time, we are able to choose "the best" (or at least pretty good) node from which to continue after we finish computing a terminal node (i.e., one with no dependents).  Current master instead uses a FIFO approach, and the previous method used a LIFO approach.  I think this is a fair tradeoff especially given the importance of this ordering.

Should we merge master into this branch before merging?

Reply:
Yeah, I think right now I agree that all of these are acceptable slowdowns.